system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,326 @@
+  return q;
+}
+
+void em_proxying_queue_destroy(em_proxying_queue* q) {
+  if (q == NULL) {
+    return;
+  }
+  assert(q != &system_proxying_queue && "cannot destroy system proxying queue");
+  // No need to acquire the lock; no one should be racing with the destruction
+  // of the queue.
+  pthread_mutex_destroy(&q->mutex);
+  for (int i = 0; i < q->size; i++) {
+    free(q->task_queues[i].tasks);
+  }
+  free(q->task_queues);
+  free(q);
+}
+
+// Not thread safe.
+static task_queue* get_tasks_for_thread(em_proxying_queue* q,
+                                        pthread_t thread) {
+  assert(q != NULL);
+  task_queue* tasks = NULL;
+  for (int i = 0; i < q->size; i++) {
+    if (pthread_equal(q->task_queues[i].thread, thread)) {
+      tasks = &q->task_queues[i];
+      break;
+    }
+  }
+  return tasks;
+}
+
+// Not thread safe.
+static task_queue* get_or_add_tasks_for_thread(em_proxying_queue* q,
+                                               pthread_t thread) {
+  task_queue* tasks = get_tasks_for_thread(q, thread);
+  if (tasks != NULL) {
+    return tasks;
+  }
+  // There were no tasks for the thread; initialize a new task_queue. If there
+  // are not enough queues, allocate more.
+  if (q->size == q->capacity) {
+    int new_capacity = q->capacity == 0 ? 1 : q->capacity * 2;
+    task_queue* new_task_queues =
+      realloc(q->task_queues, sizeof(task_queue) * new_capacity);
+    if (new_task_queues == NULL) {
+      return NULL;
+    }
+    q->task_queues = new_task_queues;
+    q->capacity = new_capacity;
+  }
+  // Initialize the next available task queue.
+  tasks = &q->task_queues[q->size];
+  tasks->thread = thread;
+  tasks->processing = 0;
+  tasks->tasks = malloc(sizeof(task) * TASK_QUEUE_INITIAL_CAPACITY);
+  if (tasks->tasks == NULL) {
+    return NULL;
+  }
+  tasks->head = 0;
+  tasks->tail = 0;
+  q->size++;
+  return tasks;
+}
+
+void emscripten_proxy_execute_queue(em_proxying_queue* q) {
+  if (q == NULL) {
+    return;
+  }
+  pthread_mutex_lock(&q->mutex);
+  task_queue* tasks = get_tasks_for_thread(q, pthread_self());
+  if (tasks == NULL || tasks->processing) {
+    // No tasks for this thread or they are already being processed.
+    pthread_mutex_unlock(&q->mutex);
+    return;
+  }
+  // Found the task queue; process the tasks.
+  tasks->processing = 1;
+  while (!task_queue_empty(tasks)) {
+    task t = task_queue_dequeue(tasks);
+    // Unlock while the task is running to allow more work to be queued in
+    // parallel.
+    pthread_mutex_unlock(&q->mutex);
+    t.func(t.arg);
+    pthread_mutex_lock(&q->mutex);
+  }
+  tasks->processing = 0;
+  pthread_mutex_unlock(&q->mutex);
+}
+
+static pthread_t normalize_thread(pthread_t thread) {
+  if (pthread_equal(thread, EM_CALLBACK_THREAD_CONTEXT_MAIN_BROWSER_THREAD)) {
+    return emscripten_main_browser_thread_id();
+  }
+  if (pthread_equal(thread, EM_CALLBACK_THREAD_CONTEXT_CALLING_THREAD)) {
+    return pthread_self();
+  }
+  return thread;
```

Comment:
Is this needed here?  Maybe we can limit the use of these macros to the high layers?

Reply:
I'm slightly worried that this will be a footgun for users, but I'm not sure how often these macros are used in the first place. I think you're right that keeping these out of the low-level API makes sense. FWIW, I decided against using `EMSCRIPTEN_RESULT_{SUCCESS,FAILURE}` here for largely the same reason; those macros also come from html5.h and seemed better suited for higher-level APIs.